src/Raven.Server/Web/RequestHandler.cs

```diff
@@ -353,12 +358,47 @@ private static void InvalidCountOfValues(string name)
             throw new ArgumentException($"Query string value '{name}' must appear exactly once");
         }
 
-        protected DisposableAction TrackRequestTime()
+        protected DisposableAction TrackRequestTime(bool alertThresholdExceeded = true)
         {
-            HttpContext.Response.Headers.Add(Constants.Headers.RequestTime, "1");
-            return null; // TODO [ppekrol] we cannot write Headers after response have started without buffering
-            //var sw = Stopwatch.StartNew();
-            //return new DisposableAction(() => HttpContext.Response.Headers.Add(Constants.Headers.RequestTime, sw.ElapsedMilliseconds.ToString(CultureInfo.InvariantCulture)));
+            var sw = Stopwatch.StartNew();
+            _context.HttpContext.Response.OnStarting(state => 
+            {
+                var httpContext = (HttpContext)state;
+                httpContext.Response.Headers.Add(Constants.Headers.RequestTime, sw.ElapsedMilliseconds.ToString());
+                return Task.FromResult(0);
+            }, _context.HttpContext);
+
+            return new DisposableAction(() =>
+            {
+                try
+                {
+                    if (alertThresholdExceeded)
+                    {
+                        var threshold = _context.RavenServer.Configuration.Core.TooLongRequestThreshold.AsTimeSpan;
+                        if (threshold.TotalMilliseconds >= sw.ElapsedMilliseconds)
+                        {
+
+                            var alert = AlertRaised.Create(
+                                "Request Latency",
+                                $@"Request latency has surpassed the configured threshold (The threshold is {threshold.TotalSeconds} sec, 
+                            but the request lasted {sw.Elapsed.TotalSeconds} sec). {Environment.NewLine} 
+                            The request path was: {_context.HttpContext.Request.Path}",
+                                AlertType.TooLongRequestDuration,
+                                NotificationSeverity.Warning);
+
+                            _context.RavenServer.ServerStore.NotificationCenter.Add(alert);
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    //precaution - should never arrive here
+                    if (_logger.IsInfoEnabled)
+                    {
+                        _logger.Info($"Failed to write request time in response headers. This is not supposed to happen and is probably a bug. The request path was: {_context.HttpContext.Request.Path}",e);
```

Comment:
This isn't what is going on here at all. This is actually the error if we can't raise the alert.